rekall-core/rekall/plugins/addrspaces/vmi.py

```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
         url = filename or (session and session.GetParameter("filename"))
         self.as_assert(url, "Filename must be specified in session (e.g. "
                        "session.SetParameter('filename', 'vmi://domain').")
-        self.as_assert(url.startswith(URL_PREFIX),
-                       "The domain must be passed with the URL prefix {}".format(URL_PREFIX))
-        domain = url[len(URL_PREFIX):]
-        self.as_assert(domain, "domain name missing after {}".format(URL_PREFIX))
+        pattern = '{}((?P<hypervisor>xen|kvm)/)?(?P<domain>.*)'.format(URL_PREFIX)
+        m = re.match(pattern, url)
+        self.as_assert(m,
+                "VMI URL format: {}(xen|kvm)?<domain>".format(URL_PREFIX))
+        domain = m.group('domain')
+        hypervisor = m.group('hypervisor')
+        self.as_assert(domain, "no domain name specified")
+        self.mode = None
+        if hypervisor:
+            self.mode = VMIMode[hypervisor.upper()]
 
         super(VMIAddressSpace, self).__init__(base=base, session=session, **kwargs)
-        self.vmi = Libvmi(domain, partial=True)
+        self.vmi = Libvmi(domain, mode=self.mode, partial=True)
+        self.min_addr = 0
+        self.max_addr = self.vmi.get_memsize() - 1
 
     def close(self):
         self.vmi.destroy()
 
     def read(self, addr, size):
-        buffer, bytes_read = self.vmi.read_pa(addr, size)
-        if bytes_read != size:
-            raise RuntimeError('Error while reading physical memory at {}'.format(hex(addr)))
+        try:
+            buffer, bytes_read = self.vmi.read_pa(addr, size)
+        except LibvmiError:
+            buffer = bytes(size)
```

Comment:
Please use the zeroer to make this more efficient.

https://github.com/google/rekall/blob/master/rekall-core/rekall/addrspace.py#L55